pkg/dataobj/sections/pointers/builder.go

```diff
@@ -0,0 +1,377 @@
+package pointers
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/loki/v3/pkg/dataobj"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/dataset"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/datasetmd"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/pointersmd"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/streamio"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/util/sliceclear"
+)
+
+// A ObjPointer is a pointer to an section within another object.
+// It is a wide object containing two types of index information:
+//
+// 1. Stream indexing metadata
+// 2. Column indexing metadata
```

Comment:
Hm, I do find it a little confusing that a pointer is to a stream _or_ a column. I think there's a few options to make that clearer:

* Store a column indicating the type of row (STREAM_POINTER vs COLUMN_POINTER), or 
* Expose a method on ObjPointer to compute the type of row (StreamPointer vs ColumnPointer)
* Make it clearer in the comment here that the type of index information is either/or per row and never both 
* Split out into two section for stream pointers and another for column pointers

Reply:
I opted for "Store a column indicating the type of row (STREAM_POINTER vs COLUMN_POINTER)"
I'm not sure how much clearer it made it: A follow up PR will be actually using this section in a builder so I might revisit it there if I think of a better way. 
The only real reason to have a single section is to reduce duplicate code and boilerplate. If extracting all the duplicated code reduces that, I doubt there is much harm in having different sections for different kinds of index.